src/rln_contract.ts

```diff
@@ -109,3 +109,85 @@ export class RLNContract {
     return txRegisterReceipt?.events?.[0];
   }
 }
+
+type CustomQueryOptions = {
+  fromBlock?: number;
+  membersFilter: ethers.EventFilter;
+};
+
+// these value should be tested on other networks
+const FETCH_CHUNK = 5;
+const BLOCK_RANGE = 3000;
+
+async function queryFilter(
+  contract: ethers.Contract,
+  options: CustomQueryOptions
+): Promise<ethers.Event[]> {
+  const { fromBlock, membersFilter } = options;
+
+  if (!fromBlock) {
+    return contract.queryFilter(membersFilter);
+  }
+
+  if (!contract.signer.provider) {
+    throw Error("No provider found on the contract's signer.");
+  }
+
+  const toBlock = await contract.signer.provider.getBlockNumber();
+
+  if (toBlock - fromBlock < BLOCK_RANGE) {
+    return contract.queryFilter(membersFilter);
+  }
+
+  const events: ethers.Event[][] = [];
+  const chunks = splitToChunks(fromBlock, toBlock, BLOCK_RANGE);
+
+  for (const portion of takeN<[number, number]>(chunks, FETCH_CHUNK)) {
+    const promises = portion.map(([left, right]) =>
+      ignoreErrors(contract.queryFilter(membersFilter, left, right), [])
+    );
+    const fetchedEvents = await Promise.all(promises);
+    events.push(fetchedEvents.flatMap((v) => v));
+  }
+
+  return events.flatMap((v) => v);
+}
+
+function splitToChunks(
+  from: number,
+  to: number,
+  step: number
+): Array<[number, number]> {
+  const chunks = [];
+
+  let left = from;
+  while (left < to) {
+    const right = left + step < to ? left + step : to;
+
+    chunks.push([left, right] as [number, number]);
+
+    left = right;
+  }
+
+  return chunks;
+}
+
+function* takeN<T>(array: T[], size: number): Iterable<T[]> {
+  let start = 0;
+  let skip = size;
+
+  while (skip < array.length) {
+    const portion = array.slice(start, skip);
+
+    yield portion;
+
+    start = skip;
+    skip += size;
+  }
+}
+
+function ignoreErrors<T>(promise: Promise<T>, defaultValue: T): Promise<T> {
+  return promise.catch((_) => {
+    return defaultValue;
```

Comment:
This is necessary because in case there are too many requests at the same time (optimal value is `FETCH_CHUNK`) then `ethers/metamask` throws an error which is not an error.

![image](https://user-images.githubusercontent.com/118575614/234407504-c922590e-1937-4e4a-b827-77c0d32a06df.png)

I think the most reasonable way is to ignore it and not populate members array.
Not sure about logging.
In case of the example it tries to fetch at this moment blocks in the range from `7109391` to `8891301`. 

So in short it requests a lot of block and there will be a lot of spamming in console if someone decides to log every error.

Reply:
Re logging maybe have a custom text that described to a potential developer the issue and possible fix?

Reply:
Added logging